### DISCUSSION TITLE
custom row classes with complex data

### DISCUSSION THREAD

**Discussion by Hanibachi** [simolus3/drift#3269] [Q&A]
i have a class and one of the fields is an object of another class , such as this example :
```dart 
class Object {
  final int id;
  final string name;
  final objectType type;
  const Object({required this.id,required this.name,required this.type});
}

class ObjectType {
  final int id;
  final string name;
  const ObjectType({required this.id,required this.name});
}
```

i want to use them as custom row classes while having two seperate tables, one for objects, the other for objectTypes . 
if it's not possible, the how about making them insertable into one table ? i don't see a way of implementing type converters in this scenario.

ps: i would appreciate if u had alternative suggestions about organizing data that might have such structure.

**Comment by simolus3** [comment]
> i want to use them as custom row classes while having two seperate tables, one for objects, the other for objectTypes .

This is indeed not possible. Row classes always represent _a single row_, and if you have two tables, you'll have at least two rows involved. With the [manager APIs](https://drift.simonbinder.eu/docs/manager/), there's a pretty straightforward way to resolve joins. So you could keep the two separate classes and then return a third class (or a record) combining `Object` and `ObjectType`.

>  if it's not possible, the how about making them insertable into one table ? i don't see a way of implementing type converters in this scenario.

In this case, you could turn the inner type (`ObjectType`) into a JSON structure:

```dart

class ObjectType {
  final int id;
  final String name;
  const ObjectType({required this.id,required this.name});

  factory ObjectType.fromJson(json) => ObjectType(json['id] as int, json['name'] as String);
  Map<String, Object> toJson() => {'id':  id, 'name': name};
}
```

And then define the table for `Object` like this:

```dart
@UseRowClass(Object)
class Objects extends Table {
  IntColumn get id => integer().autoIncrement()()
  // ...
  TextColumn get type => text().map(TypeConverter.json(fromJson: ObjectType.fromJson, toJson: (o) => o.toJson()))();
}
```
